xls/dslx/type_system/type_info.cc

```diff
@@ -155,7 +155,10 @@ absl::StatusOr<TypeInfo*> TypeInfoOwner::GetRootTypeInfo(const Module* module) {
 // -- class TypeInfo
 
 void TypeInfo::NoteConstExpr(const AstNode* const_expr, InterpValue value) {
-  const_exprs_.insert({const_expr, value});
+  VLOG(5) << absl::StreamFormat(
+      "noting node: `%s` (%p) has constexpr value: `%s`",
+      const_expr->ToString(), const_expr, value.ToString());
+  const_exprs_.insert_or_assign(const_expr, std::move(value));
```

Comment:
FYI, re review delay: The replacement of the const expr value sounds benign, however it breaks some other code outside this repository which I have yet to narrow down (time-slicing myself this week).

Highlevel: apparently in some for-loop in a proc, `NoteConstExpr()` is called with some array as value, and later for the same `AstNode` with a shorter array ... then things go south when some place attempts to access a non-existent index on the shorter array...
I've yet to determine the details. Will write a reproduction and fix when time permits.

Reply:
Alright. The new https://github.com/google/xls/blob/main/xls/examples/memory_proc.x serves as a useful example for beginners, but it will also be a repo.

Essentially, the state, which is an array representing the 'memory' is replaced with a single value. Which might actually might be different problem somewhere else.

Reply:
Ok found the issue, it was for constexpr evaluation of arrays with ellipsis.

There were cases where we were confusing the number of members in the array literal with the array size even though there were ellipses.

Also removed accidental complexity in the course of making it consistent -- now ConstexprEvaluator calls into the bytecode intepreter directly instead of reimplementing another interpretation routine.

Reply:
Forgot to tag you @hzeller -- thanks for the reproducer!